### DISCUSSION TITLE
Optimize Query costs for BigQuery in "Preview" feature

### DISCUSSION THREAD

**Discussion by fernandofernandezgonzalez** [apache/superset#13413] [Ideas]
Hi All,

First of all, I apollogize if the issue has been previously discussed, since I can't find any related issue. 
I don't know either if I should start this as a discussion or an issue.
We have been analyzing BigQuery costs in our company which has skyrocketed in the last month. We have started to make intense use of SQLLab by a few new users. Analyzing BigQuery logs we have discovered that whenever a data "Preview" is launched in SQLLab, a "Select * from table limit 100" (With no partition filters) is launched!!. This, for tables around 3TB means 3TB will be read in BigQuery, incurring in 3TB of cost!.
As per Bigquery documentation, even if limit 100 is used the whole table will be read.
We can cut cost increase by imposing "required partition filter", but this makes us lose the "Preview" feature because the query underneath fails. 
Have you guys encountered such problem? Which other techniques have you used for controlling this? Should I open an issue?

**Comment by srinify** [comment]
@junlincc @betodealmeida some interesting side effects of getting users closer to seeing / internalizing the data, more pressure on the db. 

- Caching data preview could help here
- Potential for feature flags to turn off this feature entirely, at the cost to the user
- @fernandofernandezgonzalez are there any BigQuery specific ways to minimize the amount of data scanned here? I know the new engine they're promoting / pushing is supposed to be much smarter about using caching but I don't know the details!

**Reply by 3jerde** [reply]
I have the same issues. Some wide nested tables that get expensive to do limit 100 previews on.

The bigquery python command line tool has a head command that returns a preview without query cost.

bq head schema.table

underlying this uses the API function tabledata.list

I'm guessing this would require developments in pybigquery for superset to use.

**Reply by betodealmeida** [reply]
Interesting. Leveraging this command might not be easy, since we need to go through the SQLAlchemy layer, but there might be a workaround.

Worst case we should definitely just query the latest partition. I wasn't away that BQ will scan everything for an unordered `LIMIT 100` query.

**Comment by riordan** [comment]
This would be great to combine with something like native partition awareness in SQL Lab (see: #17299).

I'd _happily_ pay Preset $20/mo for everyone in my organization if it helped my users write responsible BQ SQL in the SQLLab and smartly cost-optimized Exploration. BigQuery makes it easy for one curious user to incur massive costs on a large table and the easiest way to enable exploration while keeping costs constrained is often in the customer UI.

**Reply by betodealmeida** [reply]
Yeah, there's a lot of room for improvement here. The easiest would be passing the partition predicate, since SQL Lab has context on the partitions. We should also integrate BQ's query cost estimation with the Superset API for that.

**Reply by riordan** [reply]
@betodealmeida YOU’RE A MIND READER: #17891!

**Comment by fernandofernandezgonzalez** [comment]
Hi!

One colleague has make a pull request for the estimation costs feature for bigquery. Any chance

https://github.com/apache/superset/pull/21325

I think this will be a very appreciated feature!!
